cg/apps/demultiplex/sample_sheet/index.py

```diff
@@ -0,0 +1,126 @@
+"""Functions that deals with modifications of the indexes"""
+import csv
+import logging
+from typing import Dict, List, Set, Tuple
+
+from cg.apps.lims.samplesheet import LimsFlowcellSample
+from cg.models.demultiplex.valid_indexes import Index
+from cg.resources import valid_indexes_path
+
+LOG = logging.getLogger(__name__)
+NEW_CONTROL_SOFTWARE_VERSION = "1.7.0"
+NEW_REAGENT_KIT_VERSION = 1.5
+DNA_COMPLEMENTS = {"A": "T", "C": "G", "G": "C", "T": "A"}
+
+
+def index_exists(index: str, indexes: Set[str]) -> bool:
+    """ Determines if a index is already present in the existing indexes """
+    return any(existing_index.startswith(index) for existing_index in indexes)
+
+
+def get_indexes_by_lane(samples: List[LimsFlowcellSample]) -> Dict[int, Set[str]]:
+    """Group the indexes from samples by lane"""
+    indexes_by_lane = {}
+    for sample in samples:
+        lane: int = sample.lane
+        if lane not in indexes_by_lane:
+            indexes_by_lane[lane] = set()
+        indexes_by_lane[lane].add(sample.index)
+    return indexes_by_lane
+
+
+def get_valid_indexes(dual_indexes_only: bool = True) -> List[Index]:
+    LOG.info("Fetch valid indexes from %s", valid_indexes_path)
+    indexes: List[Index] = []
+    with open(valid_indexes_path, "r") as csv_file:
+        indexes_csv = csv.reader(csv_file)
+        for row in indexes_csv:
+            index_name = row[0]
+            index_sequence = row[1]
+            if dual_indexes_only and not is_dual_index(index_sequence):
+                continue
+            indexes.append(Index(name=index_name, sequence=index_sequence))
+    return indexes
+
+
+def get_reagent_kit_version(reagent_kit_version: str) -> float:
+    """ Derives the reagent kit version from the run parameters """
+    LOG.info("Converting reagent kit parameter %s to version", reagent_kit_version)
+    parameter_to_version = {"1": 1.0, "3": 1.5}
+    if reagent_kit_version not in parameter_to_version:
+        raise SyntaxError(f"Unknown reagent kit version {reagent_kit_version}")
+
+    return parameter_to_version[reagent_kit_version]
+
+
+def is_reverse_complement(control_software_version: str, reagent_kit_version_string: str) -> bool:
+    """If the run used the new NovaSeq control software version (NEW_CONTROL_SOFTWARE_VERSION)
+    and the new reagent kit version (NEW_REAGENT_KIT_VERSION) the second index should be the
+    reverse complement
+    """
+    LOG.info("Check if run is reverse complement")
+    if control_software_version != NEW_CONTROL_SOFTWARE_VERSION:
+        LOG.warning("Invalid control software version %s!", control_software_version)
```

Comment:
The older control software version is not invalid, it just doesn't require index 2 to be the reverse complement.

Reply:
changed